### PR TITLE
[WIP] Stats pruning

### PR DESCRIPTION
**Pull request by robert3005**
There's a lot of cleanup to do here but this performs pruning, however, we don't
yet make use of that pruning result

### CODE REVIEW

File: encodings/bytebool/src/lib.rs [hunk 1]
```diff
@@ -100,6 +100,15 @@ impl BoolArrayTrait for ByteBoolArray {
     fn maybe_null_slices_iter<'a>(&'a self) -> Box<dyn Iterator<Item = (usize, usize)> + 'a> {
         todo!()
     }
+
+    fn not(&self) -> Array {
```
Comment (robert3005) [inline encodings/bytebool/src/lib.rs:104]: rename to `inverse`

File: encodings/roaring/src/boolean/mod.rs [hunk 2]
```diff
@@ -104,6 +104,12 @@ impl BoolArrayTrait for RoaringBoolArray {
     fn maybe_null_slices_iter<'a>(&'a self) -> Box<dyn Iterator<Item = (usize, usize)> + 'a> {
         todo!()
     }
+
+    fn not(&self) -> Array {
```
Comment (robert3005) [inline encodings/roaring/src/boolean/mod.rs:108]: rename to `inverse`

File: encodings/runend-bool/src/array.rs [hunk 3]
```diff
@@ -113,6 +114,12 @@ impl BoolArrayTrait for RunEndBoolArray {
     fn maybe_null_slices_iter<'a>(&'a self) -> Box<dyn Iterator<Item = (usize, usize)> + 'a> {
         todo!()
     }
+
+    fn not(&self) -> Array {
```
Comment (robert3005) [inline encodings/runend-bool/src/array.rs:127]: rename to `inverse`

File: vortex-array/src/array/bool/mod.rs [hunk 4]
```diff
@@ -103,6 +103,12 @@ impl BoolArrayTrait for BoolArray {
     fn maybe_null_slices_iter<'a>(&'a self) -> Box<dyn Iterator<Item = (usize, usize)> + 'a> {
         Box::new(BitSliceIterator::new(self.buffer(), 0, self.len()))
     }
+
+    fn not(&self) -> Array {
```
Comment (robert3005) [inline vortex-array/src/array/bool/mod.rs:107]: rename to `inverse`

File: vortex-array/src/array/chunked/variants.rs [hunk 5]
```diff
@@ -54,6 +54,16 @@ impl BoolArrayTrait for ChunkedArray {
     fn maybe_null_slices_iter(&self) -> Box<dyn Iterator<Item = (usize, usize)>> {
         todo!()
     }
+
+    fn not(&self) -> Array {
```
Comment (robert3005) [inline vortex-array/src/array/chunked/variants.rs:58]: rename to `inverse`

File: vortex-array/src/array/constant/variants.rs [hunk 6]
```diff
@@ -78,6 +78,18 @@ impl BoolArrayTrait for ConstantArray {
             Box::new(iter::empty())
         }
     }
+
+    fn not(&self) -> Array {
```
Comment (robert3005) [inline vortex-array/src/array/constant/variants.rs:82]: rename to `inverse`

File: vortex-array/src/array/sparse/variants.rs [hunk 7]
```diff
@@ -55,6 +55,18 @@ impl BoolArrayTrait for SparseArray {
     fn maybe_null_slices_iter(&self) -> Box<dyn Iterator<Item = (usize, usize)>> {
         todo!()
     }
+
+    fn not(&self) -> Array {
```
Comment (robert3005) [inline vortex-array/src/array/sparse/variants.rs:59]: rename to `inverse`

File: vortex-array/src/array/constant/variants.rs [hunk 8]
```diff
@@ -107,20 +119,6 @@ where
 }
 
 impl PrimitiveArrayTrait for ConstantArray {
-    fn f32_accessor(&self) -> Option<AccessorRef<f32>> {
```
Comment (robert3005) [inline vortex-array/src/array/constant/variants.rs:110]: undo this diff

File: vortex-array/src/compute/compare.rs [hunk 9]
```diff
@@ -95,7 +95,7 @@ pub fn compare(
         vortex_bail!("Compare operations only support arrays of the same type");
     }
 
-    if left.is_encoding(Constant::ID) {
+    if left.is_encoding(Constant::ID) && !right.is_encoding(Constant::ID) {
```
Comment (robert3005) [inline vortex-array/src/compute/compare.rs:98]: make a separate pr and add tests

File: vortex-array/src/variants.rs [hunk 10]
```diff
@@ -103,6 +103,8 @@ pub trait BoolArrayTrait: ArrayTrait {
     //                                         and the buffer is a slice of that array, omitted slices
     //                                         could be either true or false signified by the initial
     //                                         value returned.
+
+    fn not(&self) -> Array;
```
Comment (robert3005) [inline vortex-array/src/variants.rs:107]: rename to `inverse`

File: vortex-expr/src/expr.rs [hunk 11]
```diff
@@ -6,26 +6,23 @@ use std::sync::Arc;
...
-const COLUMN_COST_MULTIPLIER: usize = 1024;
-
 pub trait VortexExpr: Debug + Send + Sync + PartialEq<dyn Any> {
     fn as_any(&self) -> &dyn Any;
 
     fn evaluate(&self, batch: &Array) -> VortexResult<Array>;
 
     fn references(&self) -> HashSet<Field>;
 
-    fn estimate_cost(&self, schema: &Schema) -> usize;
```
Comment (robert3005) [inline vortex-expr/src/expr.rs:28]: This has to come back at some point, however, it has to have different form

File: vortex-expr/src/expr.rs [hunk 12]
```diff
@@ -43,9 +40,6 @@ fn unbox_any(any: &dyn Any) -> &dyn Any {
     }
 }
 
-#[derive(Debug, PartialEq, Hash, Clone, Eq)]
-pub struct NoOp;
```
Comment (robert3005) [inline vortex-expr/src/expr.rs:47]: There was no need to translate it, we will never see it as a datafusion table provider

File: vortex-expr/src/expr.rs [hunk 13]
```diff
@@ -71,6 +65,85 @@ impl BinaryExpr {
...
+                if let Some(rhsp) = rhs_proj {
+                    Some(Arc::new(BinaryExpr::new(lhsp, self.operator, rhsp)))
+                } else {
+                    // TODO(robert): This might be too broad of a check since it should be limited only to fields in the projection
+                    self.rhs
+                        .references()
+                        .intersection(&lhsp.references())
+                        .next()
+                        .is_none()
+                        .then_some(lhsp)
```
Comment (robert3005) [inline vortex-expr/src/expr.rs:111]: Narrow this check down

File: vortex-expr/src/expr.rs [hunk 14]
```diff
@@ -6,26 +6,23 @@ use std::sync::Arc;
...
     fn as_any(&self) -> &dyn Any;
 
     fn evaluate(&self, batch: &Array) -> VortexResult<Array>;
 
     fn references(&self) -> HashSet<Field>;
 
-    fn estimate_cost(&self, schema: &Schema) -> usize;
+    fn project(&self, projection: &[Field]) -> Option<Arc<dyn VortexExpr>>;
+
+    fn is_constant(&self) -> bool;
```
Comment (robert3005) [inline vortex-expr/src/expr.rs:25]: This should actually be `fn fold(&self) -> Option<Scalar>`, i.e. can this be evaluated to a constant?

File: vortex-expr/src/lib.rs [hunk 15]
```diff
@@ -10,20 +8,25 @@ pub use expr::*;
 pub use operators::*;
 
 pub fn split_conjunction(expr: &Arc<dyn VortexExpr>) -> Vec<Arc<dyn VortexExpr>> {
```
Comment (robert3005) [inline vortex-expr/src/lib.rs:10]: Split this diff to separate pr

File: vortex-flatbuffers/flatbuffers/vortex-serde/footer.fbs [hunk 16]
```diff
@@ -4,8 +4,9 @@ table FlatLayout {
 }
 
 table NestedLayout {
-    children: [Layout];
     encoding: uint16;
+    metadata: [ubyte];
```
Comment (robert3005) [inline vortex-flatbuffers/flatbuffers/vortex-serde/footer.fbs:8]: actually use this, it's never written for now

File: vortex-scalar/src/primitive.rs [hunk 17]
```diff
@@ -171,14 +171,6 @@ macro_rules! primitive_scalar {
             }
         }
 
-        impl From<Option<$T>> for ScalarValue {
```
Comment (robert3005) [inline vortex-scalar/src/primitive.rs:174]: we can actually have a generic instead of generating all of the impls

File: vortex-scalar/src/value.rs [hunk 18]
```diff
@@ -135,6 +135,12 @@ impl ScalarValue {
     }
 }
 
+impl<T: Into<ScalarValue>> From<Option<T>> for ScalarValue {
```
Comment (robert3005) [inline vortex-scalar/src/value.rs:138]: like this

File: vortex-serde/Cargo.toml [hunk 19]
```diff
@@ -34,22 +35,22 @@ vortex-scalar = { workspace = true, features = ["flatbuffers"] }
...
+rstest = { workspace = true }
 simplelog = { workspace = true }
 tokio = { workspace = true, features = ["full"] }
 vortex-alp = { path = "../encodings/alp" }
 vortex-fastlanes = { path = "../encodings/fastlanes" }
 vortex-sampling-compressor = { path = "../vortex-sampling-compressor" }
 
 [lints]
-workspace = true
+# workspace = true
```
Comment (robert3005) [inline vortex-serde/Cargo.toml]: bring this back

File: vortex-serde/src/layouts/pruning.rs [hunk 20]
```diff
@@ -1,41 +1,91 @@
...
 impl PruningPredicate {
-    pub fn new(original_expr: &Arc<dyn VortexExpr>) -> Self {
-        let (expr, stats_to_fetch) = convert_to_pruning_expression(original_expr);
-        Self {
-            expr,
-            stats_to_fetch,
+    pub fn try_new(original_expr: &Arc<dyn VortexExpr>) -> Option<Self> {
+        let (expr, required_stats) = convert_to_pruning_expression(original_expr);
+        // TODO(robert): Could be constant false but right now we don't generate such expressions, need to have way to reduce constant expressions
+        if expr.is_constant() {
```
Comment (robert3005) [inline vortex-serde/src/layouts/pruning.rs:20]: This should use `fold`

File: vortex-serde/src/layouts/pruning.rs [hunk 21]
```diff
@@ -1,41 +1,91 @@
...
+                    )))) as Arc<dyn VortexExpr>,
+                    references,
+                )
+            })
+            .unwrap_or_else(|| {
+                (
+                    Arc::new(Literal::new(Scalar::bool(false, Nullability::NonNullable))),
+                    HashMap::new(),
+                )
+            });
```
Comment (robert3005) [inline vortex-serde/src/layouts/pruning.rs:88]: Extract this to separate function

File: vortex-serde/src/layouts/read/batch.rs [hunk 22]
```diff
@@ -53,14 +60,118 @@ impl BatchReader {
...
             Ok(Some(ReadResult::ReadMore(messages)))
         }
     }
 }
+
+#[derive(Debug)]
+pub struct BatchPruner {
+    children: Vec<Box<dyn LayoutReader>>,
+    selectors: Vec<Option<RowSelector>>,
+    projected_scans: Vec<(usize, PruningScan)>,
```
Comment (robert3005) [inline vortex-serde/src/layouts/read/batch.rs:80]: This is a bit too similar to BatchReader. Consider unifying

File: vortex-serde/src/layouts/read/builder.rs [hunk 23]
```diff
@@ -47,13 +50,8 @@ impl<R: VortexReadAt> LayoutReaderBuilder<R> {
...
     }
 
-    pub fn with_indices(mut self, array: Array) -> Self {
-        // TODO(#441): Allow providing boolean masks
-        assert!(
-            array.dtype().is_int(),
-            "Mask arrays have to be integer arrays"
-        );
-        self.indices = Some(array);
+    pub fn with_row_selector(mut self, array: RowSelector) -> Self {
```
Comment (robert3005) [inline vortex-serde/src/layouts/read/builder.rs:53]: It's strictly more powerful to use row ranges than individual indices

File: vortex-serde/src/layouts/read/builder.rs [hunk 24]
```diff
@@ -110,25 +114,54 @@ impl<R: VortexReadAt> LayoutReaderBuilder<R> {
...
+                    message_cache.clone(),
+                    pscan,
+                ))
+            } else {
+                Err((self.reader, filter_reader))
+            }
+        } else {
+            Err((self.reader, filter_reader))
+        };
+
```
Comment (robert3005) [inline vortex-serde/src/layouts/read/builder.rs:161]: This is terrible, we have to push filtering in to readers to avoid this mess

File: vortex-serde/src/layouts/read/cache.rs [hunk 25]
```diff
@@ -78,12 +88,17 @@ impl RelativeLayoutCache {
             .remove(&self.absolute_id(path))
     }
 
-    pub fn dtype(&self) -> DType {
-        self.dtype.clone()
+    pub fn has_dtype(&self) -> bool {
+        self.dtype.is_some()
+    }
```
Comment (robert3005) [inline vortex-serde/src/layouts/read/cache.rs:93]: A bit of a hack. We know that if we don't have dtype it must be stored somehow. However, I think we can deduce stats schema and avoid storing dtype altogether if we store stats as column

File: vortex-serde/src/layouts/read/layouts/chunked.rs [hunk 26]
```diff
@@ -0,0 +1,352 @@
...
+                                .with_dyn(|a| {
+                                    a.as_struct_array()
+                                        .ok_or_else(|| vortex_err!("Stats weren't a struct array"))
+                                        .map(|s| {
+                                            s.field_by_name(&format!("{prefix_str}_row_offset"))
+                                        })
+                                })?
+                                .ok_or_else(|| vortex_err!("Missing row offsets"))?;
+                            filter_offsets(row_offsets, mask, scan.row_count as usize)
+                                .map(|s| Some(PlanResult::Range(s)))
```
Comment (robert3005) [inline vortex-serde/src/layouts/read/layouts/chunked.rs:192]: There is too much logic here

File: vortex-serde/src/layouts/read/layouts/chunked.rs [hunk 27]
```diff
@@ -0,0 +1,352 @@
...
+                            let row_offsets = expanded_batch
+                                .with_dyn(|a| {
+                                    a.as_struct_array()
+                                        .ok_or_else(|| vortex_err!("Stats weren't a struct array"))
+                                        .map(|s| {
+                                            s.field_by_name(&format!("{prefix_str}_row_offset"))
+                                        })
+                                })?
+                                .ok_or_else(|| vortex_err!("Missing row offsets"))?;
+                            filter_offsets(row_offsets, mask, scan.row_count as usize)
```
Comment (robert3005) [inline vortex-serde/src/layouts/read/layouts/chunked.rs:191]: Since we return chunks that should be pruned and we want to produce rows to include we need to flip the mask in this function

File: vortex-serde/src/layouts/read/layouts/chunked.rs [hunk 28]
```diff
@@ -0,0 +1,352 @@
...
+                    },
+                    self.message_cache.relative_stored_dtype(0u16),
+                )?,
+            );
+            self.plan(scan)
+        }
+    }
+}
+
+fn filter_offsets(offsets: Array, mask: Array, row_count: usize) -> VortexResult<RowSelector> {
```
Comment (robert3005) [inline vortex-serde/src/layouts/read/layouts/chunked.rs:229]: The mask has to be flipped before being passed to this function or this function flips the mask

File: vortex-serde/src/layouts/read/layouts/flat.rs [hunk 29]
```diff
@@ -0,0 +1,97 @@
...
+                    let dtype_buf = buf.split_to(dtype_fb_length as usize);
+                    let msg = unsafe { root_unchecked::<fb::Message>(&dtype_buf) }
+                        .header_as_schema()
+                        .ok_or_else(|| {
+                            vortex_err!(
+                                "Expected schema message; this was checked earlier in the function"
+                            )
+                        })?;
+                    IPCDType::read_flatbuffer(&msg)?.0
+                };
```
Comment (robert3005) [inline vortex-serde/src/layouts/read/layouts/flat.rs:78]: this likely goes away if we change the way stats are stored

File: vortex-serde/src/layouts/read/stream.rs [hunk 30]
```diff
@@ -16,44 +17,57 @@ use vortex_schema::Schema;
...
+            layout_pruner = Some(pruner);
+            state = StreamingState::Pruning;
         } else {
-            StreamingState::Init
-        };
+            (input, filter_reader) = pruner.map_err(|(i, f)| (Some(i), f)).unwrap_err();
+            if filter_reader.is_some() {
+                state = StreamingState::FilterInit;
+            }
+        }
```
Comment (robert3005) [inline vortex-serde/src/layouts/read/stream.rs]: I hate this code but ownership is hard

File: vortex-serde/src/layouts/read/selection.rs [hunk 31]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+/// Sorted list of row ranges to be included when reading vortex files
+#[derive(Debug, Clone, Default, PartialEq, Eq)]
+pub struct RowSelector {
+    ranges: Vec<RowRange>,
+}
+
+impl RowSelector {
+    pub fn new(ranges: Vec<RowRange>) -> Self {
```
Comment (robert3005) [inline vortex-serde/src/layouts/read/selection.rs:37]: We should have a sorting constructor and unsafe non sorting one

File: encodings/bytebool/src/lib.rs (hunk 1 above)
Comment (gatesn) [reply encodings/bytebool/src/lib.rs:104]: Invert? Negate?

**Comment by robert3005** [comment]
This pr needs to be split in 3
1. Add Not Expression (and implementations)
2. Add methods for adding missing columns
3. Read metadata and run pruning
